### ISSUE TITLE
`npm config get prefix` takes incredibly long (7 - 70 seconds)

### ISSUE THREAD

**Issue by polybuildr** [npm/npm#14458]
#### I'm opening this issue because:
- [x] Other
#### What's going wrong?

`npm config get prefix` takes incredibly long. I was chasing this issue in https://github.com/creationix/nvm/issues/1261 where `nvm` was taking incredibly long to load, and then realised that it was just the `npm config get prefix` command that was responsible for 95%+ of the slowdown.
#### How can the CLI team reproduce the problem?

Unfortunately, this is a new machine and this problem has been happening from the first time I installed npm using nvm. Fresh installs of nvm haven't changed anything, the slowdown is still there.
### supporting information:
- `npm -v` prints: 3.10.3
- `node -v` prints: v6.7.0
- `npm config get registry` prints: https://registry.npmjs.org/
- Ubuntu 16.04, Skylake i7 processor
- shell: `zsh` (but similar problems also occurred when using `bash`)

I understand that the original issue was when using `nvm`, but since it was specifically the `npm` command that took extremely long, I came here to figure out why. If there is any further profiling or debugging to do, I would love to get to the bottom of this.

**Comment by legodude17** [comment]
Can you send in a `time "npm config get prefix"`?

**Comment by polybuildr** [comment]
Another thing I forgot to mention is that this only happens the first time it is called. Basically the first time after boot, after which it's just a few hundred milliseconds for all successive calls. I'll reboot and get a `time. Also, I can `strace` the command, but I'm not sure how to do a time-based profile of it. Any suggestions? Or should I run a profiler and see what function calls are made?

**Comment by legodude17** [comment]
I think I meant `strace`. Sorry for the confusion. ☹️ Also, it is really odd that it only happens on the first time. Do you have any strange config?

**Comment by polybuildr** [comment]
Yeah, it seems rather strange. I don't think I have any strange config, in fact I don't think I have any custom config at all:

```
[vivek@matrix] ~ $ npm config list
; cli configs
user-agent = "npm/3.10.3 node/v6.7.0 linux x64"

; userconfig /home/vivek/.npmrc
v6.7.0 = true

; node bin location = /home/vivek/.nvm/versions/node/v6.7.0/bin/node
; cwd = /home/vivek
; HOME = /home/vivek
; "npm config ls -l" to show all defaults.
```

**Comment by legodude17** [comment]
Any idea why this is happening @KenanY or @othiym23 or @zkat?

**Comment by ljharb** [comment]
fwiw, `npm config get prefix` is also the slowest part of http://nvm.sh, by a long shot - if there's a faster way that's equally reliable to get the prefix setting, i'd love to use that instead.

**Comment by legodude17** [comment]
@ljharb, if you look at the top, you will see that this issue was opened from @polybuildr investigating that issue in `nvm`. Another way to do it would be `npm prefix -g`, but that might not be faster.

**Comment by ljharb** [comment]
@legodude17 yes, if you look at the linked nvm issue, you'll see that i requested they file this very issue. `npm prefix -g` is an alternative i'll look into, thanks.

**Comment by legodude17** [comment]
Speaking of `npm prefix -g`, @polybuildr, can you try `time npm config get prefix` and `time npm prefix -g`?

**Comment by zkat** [comment]
🤔 I wonder why it hangs like that.

Our startup process could use some work, and it generally looks like this is an occasional hiccup -- the bulk of the startup process is just reading config files and loading dependencies. Some of those are bigger than others and maybe not necessary, but this is worth taking a look at. 7+ seconds is kinda extreme.

**Comment by polybuildr** [comment]
Here is the output of `zsh` `time`:

https://gist.github.com/polybuildr/4e9d15508219d55b785de0b0eabe69c8

I'm not sure there is much to be gained by the comparison between `npm config get prefix` and `npm -g prefix` because even in just `npm config get prefix` I see variations between 5 seconds and 70 seconds.

I'll run an `strace` on next boot.

**Comment by legodude17** [comment]
Ok. So `npm prefix -g` doesn't help much. The `strace` output will hopefully help.

**Comment by polybuildr** [comment]
I ran `strace -r` (print relative time deltas after each command). The output is really large, but here it is:

http://paste.ubuntu.com/23400796/

I'll run another strace with the absolute time printed and I'll also try ltrace later maybe, if that will help.

**Comment by polybuildr** [comment]
And here is the output of `strace -ttt`: http://paste.ubuntu.com/23403626/

**Comment by watilde** [comment]
I've tried to sort the result of `strace -r`, but I couldn't get the point related to npm. https://gist.githubusercontent.com/watilde/0701a82acfaf8cd87658274d8a1822d2/raw/49b3e3a11fca5496387c582254fe3e461bc6b822/gistfile1.txt

I was considering if the bootstrap makes any HTTP-request, and I could see some of the potentials of that from the following codes, but the above log didn't suggest any slow HTTP-request.

e.g.
1. https://github.com/npm/npm/blob/latest/bin/npm-cli.js#L29
2. https://github.com/npm/npm/blob/latest/lib/npm.js#L32
3. https://github.com/npm/npm/blob/latest/lib/cache/caching-client.js#L9
4. https://github.com/npm/npm-registry-client/blob/master/index.js#L73
5. https://github.com/npm/npm-registry-client/blob/master/lib/fetch.js#L90

**Comment by martinheidegger** [comment]
I have been looking into the nvm issue as well in order to figure out why the nvm is so slow to start. In my research I figured that npm is the slowest part of npm. 

On my machine nvm takes almost 3 seconds to start up. I wrote some old-school tests for the performance and they showed that most operations are quick but for a few tasks it has to use npm which makes it slow. 

https://github.com/martinheidegger/nvm/tree/debug/performance

_(It writes the results in the `timings.txt` in the $NVM_DIR)_

Particularly interesting is npm config get prefix which - once called - consumes 0.5 seconds itself in the test environment. To make sure that its not a recent issue I wrote some performance tests that came up with the avg 0.5 seconds:
https://gist.github.com/martinheidegger/32d00e90e0163a22a4ffc78df796001e

**Comment by martinheidegger** [comment]
Edit: I updated [the tests](https://gist.github.com/martinheidegger/32d00e90e0163a22a4ffc78df796001e) by running [a hack version of `npm config`](https://gist.github.com/martinheidegger/32d00e90e0163a22a4ffc78df796001e#file-npm_config_hack-sh) that excludes all the npm command line overhead and just loads the data. It seems like half of the time is spent on things other than "loading the config". Would be be reasonable to extract the `lib/config` code into a separate package that it turn could be improved for performance?

**Comment by legodude17** [comment]
There is a long term goal for npm to move into more smaller packages, but
at this point you might just want to copy lib/config.js and most of the
config folder into your own module and use that.

**Comment by ljharb** [comment]
@legodude17 if that package was created, would a PR be accepted that switched to use it? (to ensure that npm won't break that functionality in the future)

**Comment by legodude17** [comment]
I would say "no", but I am not on the CLI team, so maybe ask @othiym23?

**Comment by zkat** [comment]
It is not a high priority right now, but we've definitely been looking at what exactly is causing the slowdown, and it's a mixture of several things. Config itself isn't even half of it. The most noticeable thing is spending ~200ms on a single one of our dependencies (which, in the case of `npm prefix -g` is not necessary).

But the way that stuff is factored right now makes it fairly tricky to lazy-load: not impossible, but one of us needs to spend Actual Time on it. The rest involves a long-term project to get rid of the massive global config-object in favor of something a bit more granular, but we really haven't gotten to the point of being able to design that yet.

@ljharb I'd say to go and make that PR if it weren't a thing that required more serious architectural planning/design from the core team. I'm sure if a community member benchmarks the actual startup process and finds a minimal-impact way to cleanly and reliably lazy-load said dependencies and is willing to do a little bit of back-and-forth, we might look at a PR. But please don't do the whole thing with all the tests before talking about it with us, cause I can't guarantee we'd pull it in. That would be a nearly 50% speedup right there on startup.

I *am* concerned, though, about why it spikes to `7-70` seconds, as I can't reproduce the particular thing that @polybuildr is mentioning. If we had benchmark information about where all that time is going, we might even be able to target a more specific bit, but my best guess right now is that it's just a disk access hiccup from having to do all the work of loading said Big Dep.

**Comment by martinheidegger** [comment]
I read the code in [`lib/config/find-prefix.js`](https://github.com/npm/npm/blob/latest/lib/config/find-prefix.js) and found quite a few steps that could increase performance and those are actually logical overlaps with the Node.js code. I [asked Node.js/help](https://github.com/nodejs/help/issues/396) if & why the code is split and it seems that it would be possible to ask/propose an API to node.js to receive the prefix quicker.

> If npm wanted to use that, they should file an issue or pull request to make it public API.

Would an API like: `module.prefix` be useful to NPM?

**Comment by ljharb** [comment]
This is causing huge slowdowns in `nvm`, so it'd be appreciated if this could be left open until it's addressed.

**Comment by SamHH** [comment]
Likewise, I'm subscribed for the same reason; massive slowdowns in `nvm`.

**Comment by ljharb** [comment]
@KenanY thank you!

**Comment by ljharb** [comment]
(ping, to avoid npmbot carnage)

**Comment by ljharb** [comment]
(ping again)

**Comment by PeterDaveHello** [comment]
(help ping)

**Comment by williscool** [comment]
(Ping like the other cool kids.)

Also you can very consistently reproduce this on Windows Subsystem for Linux (WSL) aka Bash on Windows.

It happens literally every time on every shell invocation without fail.

**Comment by QuentinLemCode** [comment]
Hello,

I got this issue with npm 5.5.1 on Windows 7
NodeJs v8.9.1
![npm problem](https://user-images.githubusercontent.com/3814031/32662488-04633362-c62b-11e7-9b3d-74704d9ea83e.png)

**Comment by cristianl** [comment]
Not that another ping is needed, but just a heads-up — besides slow shell startup, this may be causing PATH bugs in other applications: https://github.com/SublimeTextIssues/Core/issues/1877

**Comment by nylen** [comment]
It looks like the long load times here are mostly explainable by filesystem reads.  Here is what I see after a fresh boot / manual cache clear on a small ARM thing similar to a Raspberry Pi:

```
$ time npm --version
5.6.0

real	0m10.582s
user	0m2.510s
sys	0m3.490s
```

Then subsequent runs of `npm --version` are much faster:

```
$ time npm --version
5.6.0

real    0m2.844s
user    0m2.545s
sys     0m0.170s
```

`npm config get prefix` is slightly slower but pretty comparable.

So npm is simply loading a lot of stuff for this low-powered system to deal with (storage reads are slow).  This looks better when system caches are primed, but it doesn't have a very powerful CPU either so we still spend about 3 seconds.

For anyone looking to address this issue within npm, it will be useful to debug the loading sequence on a Raspberry Pi.  These links should be helpful:

- [Explanation of Linux disk caching](https://www.thomas-krenn.com/en/wiki/Linux_Page_Cache_Basics)
- [More specific examples](https://www.linuxatemyram.com/play.html), including a command to drop the cache: `echo 3 | sudo tee /proc/sys/vm/drop_caches`
